[DOCS] data quality uniqueness article

Quantisan opened the pull request:
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

Quantisan committed:
uniqueness article

sets Expectations section

update sample data table

update Expectations section

add tips

fleshed out 'Example'

update Scenarios

add Pitfalls section

conclusion

add intro

copy edit Scenario

add explanation to Data preview

shorten conclusion

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md:
```diff
@@ -0,0 +1,263 @@
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+Data uniqueness is a fundamental aspect of data quality that ensures distinct values are present where expected in a dataset. Uniqueness constraints are often applied to columns that serve as primary keys, unique identifiers, or timestamps. Validating uniqueness is critical for maintaining data integrity, preventing duplication, and enabling accurate analysis.
```
```suggestion
Data uniqueness is a fundamental aspect of data quality that ensures distinct values are present only once where expected in a dataset. Uniqueness constraints are often applied to columns that serve as primary keys, unique identifiers, or timestamps. Validating uniqueness is critical for maintaining data integrity, preventing duplication, and enabling accurate analysis.
```

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md:
```diff
@@ -0,0 +1,263 @@
+:::
+
+
+### Sets-based Expectations
```
The set-based Expectations are more aligned with data validity & consistency than uniqueness - I don't think it makes sense to include these Expectations in this article. Do you see a compelling reason to keep them?

Quantisan replied:
removed 655910eec1c4bbb97831c746a36d1208518ba82e

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md:
```diff
@@ -0,0 +1,263 @@
+
+**GX solution**: Use `ExpectColumnDistinctValuesToEqualSet` to validate that the distinct values in the payment type column exactly match the set of allowed payment types. If unexpected values are found, investigate and resolve the discrepancy.
+
+### Monitoring for missing sensor readings
```
How does this scenario correspond to uniqueness?

Quantisan replied:
removed 1195d9aee259a1198ed7ccc0db2e406830c9e3ec

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md:
```diff
@@ -0,0 +1,263 @@
+
+The examples in this guide use a sample transaction dataset, available as a [CSV file on GitHub](https://raw.githubusercontent.com/great-expectations/great_expectations/develop/tests/test_sets/learn_data_quality_use_cases/uniqueness.csv).
+
+| transfer_type     | sender_account_number  | recipient_fullname | transfer_amount | transfer_ts       |
```
I think we could help readers better connect to and appreciate real world data uniqueness challenges with a customer dataset that is representative of the issues introduced in the blog post. The concept of duplicated customer records/profiles is relatable, and it would present a dataset where uniqueness could not be easily be enforced by a unique key constraint in a SQL table.

What do you think about a customer dataset with the following fields? (We can add others if you think they'll be useful for the examples):

- customer id (autoincrement)
- first name
- last name
- email address
- phone number
- country
- government id (e.g. SSN, SIN)

The sample data can then contain duplicate customers, but registered under different names, aliases, emails, etc.

Quantisan replied:
good idea! updated data and examples a8201b261ef9cec056af09783a6259365c055efc

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md:
```diff
@@ -0,0 +1,263 @@
+
+**GX solution**: Use `ExpectColumnValuesCounts` to check that the number of readings per sensor per time period (e.g., hourly) matches the expected count. If counts are lower than expected, investigate the cause of the missing readings.
+
+### Ensuring consistency in product categories
```
This seems like a validity/consistency scenario.

Quantisan replied:
removed 347a548ce3cf1399bc1ea50d0b180c03246e315e

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md:
```diff
@@ -0,0 +1,263 @@
+- **Not considering business context**: Understand specific uniqueness requirements for each dataset and use case. Blindly applying generic checks can lead to false alarms or missed issues.
+- **Checking at the wrong granularity**: Validate uniqueness at the appropriate level, whether it's individual columns or combinations of columns, based on business requirements.
+- **Mishandling missing or null values**: Decide whether to consider null values as distinct or ignore them in uniqueness validation. Be consistent to avoid skewed results.
+- **Ignoring subtle differences**: Be aware of whitespace, case sensitivity, and type mismatches that could cause false negatives. Clean and normalize data before uniqueness checks.
```
Good pitfalls!

Quantisan committed:
Update docs/docusaurus/docs/reference/learn/data_quality_use_cases/uniqueness.md

Co-authored-by: Rachel House <rachel.house@greatexpectations.io>

Quantisan committed:
remove Scenario: Monitoring for missing sensor readings

doesn't seem relevant

Quantisan committed:
removed Scenario: Ensuring consistency in product categories

Quantisan committed:
remove sets-based expectations

https://github.com/great-expectations/great_expectations/pull/10584#discussion_r1821579904

Quantisan committed:
replace sample data with customer dataset

https://github.com/great-expectations/great_expectations/pull/10584#discussion_r1821588395

Quantisan committed:
add uniqueness .py files to examples_under_test

Quantisan committed:
uniqueness_customers.csv

Quantisan committed:
fix csv link

Quantisan committed:
moved Expectations to CI

Quantisan committed:
add uniqueness_workflow.py stub

rachhouse committed:
resolve merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
get Netlify build working

rachhouse committed:
save in-progress edits

rachhouse committed:
resolve merge conflicts

rachhouse committed:
edits

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
expand sample data and gx core workflow

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
reframe example

rachhouse committed:
Merge branch 'develop' into docs/data-quality-uniqueness-2

rachhouse committed:
add referenced expectations to public api missing exceptions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
correct test copy paste

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
Merge branch 'develop' into docs/data-quality-uniqueness-2

rachhouse reviewed (approved):
No changes from internal review (conducted via Slack) - I'm approving this PR so it can get merged in.
